### DISCUSSION TITLE
Error code optimization

### DISCUSSION THREAD

**Discussion by polluks** [cc65/cc65#1794] [Ideas]
I would like to change
```
        lda     #<JOY_ERR_OK
        ldx     #>JOY_ERR_OK
```
into
```
        lda     #JOY_ERR_OK
        tax     ; x=0
```
We had a similar discussion time ago.

**Comment by spiro-trikaliotis** [comment]
Of course, this makes sense only if it can be guaranteed that JOY_ERR_OK is 0.

I have had something similar in mind, but in a more general way. For example, add a macro LDAX (or similar) which loads the low byte into A and the high byte into X. With conditional compilation, one can make sure that a TAX is used in case that low- and high-byte are equal.

But, of course, this means that these macros have to be added and used, and I think this is something that is not wanted at the moment, right?

**Comment by mrdudz** [comment]
I don't like macros for this sort of thing at all - as they hide what actually happens, make code harder to read, and make it much harder to spot obvious optimizations as a result. My rule of thumb is that a good macro is a macro that does not exist :)

That said, why not make the optimization in the code and put an assertion before that which checks if JOY_ERR_OK is 0 or not?

**Comment by spiro-trikaliotis** [comment]
> I don't like macros for this sort of thing at all

I know. ;)

> as they hide what actually happens

For such a trivial macro, this is arguable

> make code harder to read,

Once you are used to these macros, they make it even easier to read, as they also give you an intention. (Does the LDA #... just handle the Accu? Is it somehow related to the next LDX #..., or is the LDX just for a loop counter, for example?)

> and make it much harder to spot obvious optimizations as a result

Yes, this is a point. On the other hand, such a macro can also perform some simple optimizations in some cases, so it might not be that obvious what is better.

> My rule of thumb is that a good macro is a macro that does not exist :)

I know...

> That said, why not make the optimization in the code and put an assertion before that which checks if JOY_ERR_OK is 0 or not?

I do not disagree; the only thing here is, that the assertion should be there.
Of course, the assertion can be also that the high- and the low-byte of JOY_ERR_OK are the same. ;)

**Comment by mrdudz** [comment]
> Once you are used to these macros, they make it even easier to read

Sure, thats the whole point. I want code to be readable without prior knowledge - Stuff like bgt and blt should have never been created or used in this code for the same reason - its a great way to tell ppl, who dont already know the code, that you don't care about them being able to understand the code easily at all.

whatever, this case is clear to me:

- the optimization is valid and should be done (obviously)
- an error should be created if the optimization is no more valid because JOY_ERR_OK changed

doing an assertion check seems to be the logical conclusion there. i'd propose doing THAT in some sort of macro :=)
